### PR TITLE
refactor: use asyncio.run once by engine run

### PR DESCRIPTION
**Pull request by sileht**
With this change asyncio.run() is used into two places only,
at the beginning of engine thread and the simulator thread.

### CODE REVIEW

**Comment by sileht** [comment]
5 tests remaining :smile:

File: mergify_engine/rules/types.py [hunk 1]
```diff
@@ -54,7 +54,28 @@ def _ensure_complete():
         pass
 
 
-_DUMMY_PR = context.PullRequest(
+class DummaryPullRequest(context.PullRequest):
```
Comment (jd) [inline mergify_engine/rules/types.py:57]:
```suggestion
class DummyPullRequest(context.PullRequest):
```

File: mergify_engine/rules/types.py [hunk 2]
```diff
@@ -54,7 +54,28 @@ def _ensure_complete():
...
+            infos = {}
+            for k in used_variables:
+                if extra_variables and k in extra_variables:
+                    infos[k] = extra_variables[k]
+                else:
+                    infos[k] = getattr(self, k)
+            return env.from_string(template).render(**infos)
+
+
+_DUMMY_PR = DummaryPullRequest(
```
Comment (jd) [inline mergify_engine/rules/types.py:78]:
```suggestion
_DUMMY_PR = DummyPullRequest(
```

File: mergify_engine/rules/types.py [hunk 3]
```diff
@@ -54,7 +54,28 @@ def _ensure_complete():
...
+        )
+        with self._template_exceptions_mapping():
+            used_variables = jinja2.meta.find_undeclared_variables(env.parse(template))
+            infos = {}
+            for k in used_variables:
+                if extra_variables and k in extra_variables:
+                    infos[k] = extra_variables[k]
+                else:
+                    infos[k] = getattr(self, k)
+            return env.from_string(template).render(**infos)
```
Comment (jd) [inline mergify_engine/rules/types.py:75]: Is this duplicated from `context`?

**Review by jd — DISMISSED** [review]

File: mergify_engine/rules/types.py (hunk 3 above)
Comment (sileht) [reply mergify_engine/rules/types.py:75]: One is sync the other is async.

**Review by jd — APPROVED** [review]

**Pull request merged**